parseComCode/main.js

```diff
@@ -30,6 +30,44 @@ Parse.Cloud.job("populateSearch", function(request, status) {
   });
 });
 
+// This job will remove any language records which currently do not have any books which use them.
+// The purpose is to keep BloomLibrary.org from displaying languages with no books.
+// It is scheduled to run every day. (parse.com -> Core -> Jobs)
+// It can be run manually using the following from the command line:
+// curl -X POST -H "X-Parse-Application-Id: <insert app ID>"  -H "X-Parse-Master-Key: <insert master key>" -d '{}' https://api.parse.com/1/jobs/removeUnusedLanguages
+// (In theory, the -d '{}' shouldn't be needed because we are not passing any parameters, but it doesn't work without it.)
+Parse.Cloud.job("removeUnusedLanguages", function(request, status) {
+    // Set up to modify data
+    Parse.Cloud.useMasterKey();
+
+    var allLangQuery = new Parse.Query('language');
+    allLangQuery.find().then(function (languages) {
+        var promise = Parse.Promise.as();
+        for (var i = 0; i < languages.length; i++) {
+
+            // Use IIFE to pass the correct language down
+            (function() {
+                var lang = languages[i];
+                var bookQuery = new Parse.Query('books');
+                bookQuery.equalTo('langPointers', lang);
+                promise = promise.then(function () {
+                    return bookQuery.count().then(function (count) {
+                        if (count === 0) {
+                            console.log("Deleting language " + lang.get('name') + " because no books use it.");
+                            lang.destroy();
```

Comment:
Thanks.  This promise stuff is confusing...  I returned to promise to ensure it doesn't finish before success is called.